File: src/com/andrewbrookins/idea/wrap/WrapAction.java
package com.andrewbrookins.idea.wrap;

import com.intellij.openapi.actionSystem.ActionPlaces;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.actionSystem.LangDataKeys;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.SelectionModel;
import com.intellij.openapi.editor.actionSystem.EditorAction;
import com.intellij.openapi.editor.actionSystem.EditorActionHandler;
import com.intellij.openapi.project.Project;
import com.intellij.psi.codeStyle.CodeStyleSettings;
import com.intellij.psi.codeStyle.CodeStyleSettingsManager;


public class WrapAction extends EditorAction {
    private static final Logger log = Logger.getInstance(WrapAction.class);

    public WrapAction() {
        super(new WrapHandler());
    }

    @Override
    public void update(AnActionEvent e) {
        super.update(e);
        if (ActionPlaces.isPopupPlace(e.getPlace())) {
            e.getPresentation().setVisible(e.getPresentation().isEnabled());
        }
    }

    private static class WrapHandler extends EditorActionHandler {
        @Override
        public void execute(final Editor editor, final DataContext dataContext) {
            ApplicationManager.getApplication().runWriteAction(new Runnable() {
                public void run() {
                    final Project project = LangDataKeys.PROJECT.getData(dataContext);
                    final Document document = editor.getDocument();
                    final SelectionModel selectionModel = editor.getSelectionModel();
                    final int codeStyleRightMargin = editor.getSettings().getRightMargin(project);


                    if (!selectionModel.hasSelection()) {
                        selectionModel.selectLineAtCaret();
                    }

                    final String text = selectionModel.getSelectedText();
                    CodeWrapper wrapper = new CodeWrapper(codeStyleRightMargin);
                    String wrappedText = wrapper.wrap(text);

                    document.replaceString(selectionModel.getSelectionStart(),
                        selectionModel.getSelectionEnd(), wrappedText);
                }
            });
        }
    }
}

File: src/main/java/com/andrewbrookins/idea/wrap/WrapAction.kt
package com.andrewbrookins.idea.wrap

import com.intellij.openapi.actionSystem.*
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.util.TextRange
import org.jetbrains.annotations.NotNull


fun isWhitespace(str: String?): Boolean {
    if (str == null) {
        return false
    }
    val sz = str.length
    for (i in 0 until sz) {
        if (!Character.isWhitespace(str[i])) {
            return false
        }
    }
    return true
}


class WrapAction : AnAction() {
    @NotNull
    override fun getActionUpdateThread(): ActionUpdateThread {
        return ActionUpdateThread.BGT
    }

    override fun update(e: AnActionEvent) {
        e.presentation.isEnabled = true
        e.presentation.isVisible = true
    }

    override fun actionPerformed(e: AnActionEvent) {
        val dataContext = e.dataContext
        val editor = e.getData(CommonDataKeys.EDITOR) ?: return
        val project = dataContext.let { LangDataKeys.PROJECT.getData(it) }
        val fileIsPlaintext = isPlaintext(dataContext)
        val document = editor.document
        val selectionModel = editor.selectionModel
        val wrapper = getWrapper(project, editor, fileIsPlaintext)
        val text: String
        var start: Int
        val end: Int

        if (selectionModel.hasSelection()) {
            start = selectionModel.selectionStart
            end = selectionModel.selectionEnd

            // Handle the case where a user selects a line but leaves out
            // whitespace at the start of the line. We need that whitespace
            // to correctly wrap new lines in the result.
            val line = document.getLineNumber(start)
            val lineStartOffset = document.getLineStartOffset(line)
            val possibleWhitespaceEnd = if (start == 0) start else (start - 1).coerceAtLeast(lineStartOffset)
            val possibleWhitespace = document.getText(TextRange(lineStartOffset, possibleWhitespaceEnd))
            if (isWhitespace(possibleWhitespace)) {
                // The selected line has leading whitespace, so expand the selection
                // to include the whitespace.
                start = lineStartOffset
            }
        } else {
            val line = editor.caretModel.logicalPosition.line
            start = document.getLineStartOffset(line)
            end = document.getLineEndOffset(line)
        }

        text = document.getText(TextRange(start, end))
        if (text.isBlank()) {
            return
        }
        val wrappedText = wrapper.wrap(text)

        WriteCommandAction.runWriteCommandAction(project) {
            document.replaceString(start, end, wrappedText)
        }
    }
}
